cub/cub/device/dispatch/dispatch_merge.cuh

```diff
@@ -0,0 +1,366 @@
+  __shared__ typename vsmem_helper_t::static_temp_storage_t shared_temp_storage;
+  auto& temp_storage = vsmem_helper_t::get_temp_storage(shared_temp_storage, global_temp_storage);
+  MergeAgent{
+    temp_storage.Alias(),
+    make_load_iterator(MergePolicy{}, keys1),
+    make_load_iterator(MergePolicy{}, items1),
+    num_keys1,
+    make_load_iterator(MergePolicy{}, keys2),
+    make_load_iterator(MergePolicy{}, items2),
+    num_keys2,
+    keys_result,
+    items_result,
+    compare_op,
+    merge_partitions}();
+  vsmem_helper_t::discard_temp_storage(temp_storage);
+}
+
+template <typename KeyT, typename ValueT>
+struct device_merge_policy_hub
+{
+  static constexpr bool has_values = !::cuda::std::is_same<ValueT, NullType>::value;
+
+  using tune_type = char[has_values ? sizeof(KeyT) + sizeof(ValueT) : sizeof(KeyT)];
+
+  struct policy300 : ChainedPolicy<300, policy300, policy300>
+  {
+    using merge_policy =
+      agent_policy_t<128,
+                     Nominal4BItemsToItems<tune_type>(7),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_DEFAULT,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  struct policy350 : ChainedPolicy<350, policy350, policy300>
+  {
+    using merge_policy =
+      agent_policy_t<256,
+                     Nominal4BItemsToItems<tune_type>(11),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_LDG,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  struct policy520 : ChainedPolicy<520, policy520, policy350>
+  {
+    using merge_policy =
+      agent_policy_t<512,
+                     Nominal4BItemsToItems<tune_type>(13),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_LDG,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  struct policy600 : ChainedPolicy<600, policy600, policy520>
+  {
+    using merge_policy =
+      agent_policy_t<512,
+                     Nominal4BItemsToItems<tune_type>(15),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_DEFAULT,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  using max_policy = policy600;
+};
+
+template <typename KeyIt1,
+          typename ValueIt1,
+          typename KeyIt2,
+          typename ValueIt2,
+          typename KeyIt3,
+          typename ValueIt3,
+          typename Offset,
+          typename CompareOp,
+          typename PolicyHub = device_merge_policy_hub<value_t<KeyIt1>, value_t<ValueIt1>>>
+struct dispatch_t
+{
+  using key_t   = cub::detail::value_t<KeyIt1>;
+  using value_t = cub::detail::value_t<ValueIt1>;
+
+  // Cannot check output iterators, since they could be discard iterators, which do not have the right value_type
+  static_assert(::cuda::std::is_same<cub::detail::value_t<KeyIt2>, key_t>::value, "");
+  static_assert(::cuda::std::is_same<cub::detail::value_t<ValueIt2>, value_t>::value, "");
+  static_assert(::cuda::std::__invokable<CompareOp, key_t, key_t>::value,
```

Comment:
I believe those are not the right constraints and they have broken cuDF

Why do we require all keys / values to have the exact same type? Shouldnt it be sufficient to ensure that the CompareOp is invocable with the passed in types

Reply:
Because this is what the [documentation](https://nvidia.github.io/cccl/thrust/api/function_group__merging_1ga3d2776685a00dca265399f411784acec.html) of `thrust::merge` says:
```
InputIterator1 and InputIterator2 have the same value_type
```
Apparently, the iterators for the values don't need to have the same `value_type`s.